lib/Dialect/D2M/Analysis/DstAnalysisGraphColoring.cpp

```diff
@@ -0,0 +1,172 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/D2M/Analysis/DstAnalysis.h"
+#include "ttmlir/Dialect/D2M/IR/D2MGenericRegionOps.h"
+#include "ttmlir/Dialect/D2M/IR/D2MOps.h"
+#include "ttmlir/Dialect/D2M/Transforms/GraphColoringStrategy.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+
+namespace mlir::tt::d2m {
+
+namespace {
+
+/// Identify affine load and store operations that require DST allocation.
+/// Uses getOperandsLoadFromDstRegister() to identify loads and
+/// getDstRegInPlace() to determine if stores need separate DST slices.
+static llvm::SmallVector<std::pair<mlir::Operation *, int64_t>>
+identifyDstAccesses(mlir::Operation *op, mlir::Region &region) {
+  llvm::SmallVector<std::pair<mlir::Operation *, int64_t>> dstAccesses;
+  int64_t index = 0;
+
+  // Walk the region looking for D2M compute operations
+  region.walk([&](mlir::Operation *operation) {
+    // Check if this is a D2M compute operation with DST interface
+    auto computeOp =
+        llvm::dyn_cast<OperandLoadStoreRegisterOpInterface>(operation);
+    if (!computeOp) {
+      return;
+    }
+
+    // Get operand indices that must be loaded from DST register
+    auto operandsLoadFromDst = computeOp.getOperandsLoadFromDstRegister();
+
+    // For each operand that loads from DST, find the corresponding load
+    // operation
+    for (int64_t operandIdx : operandsLoadFromDst) {
+      mlir::Value operand = operation->getOperand(operandIdx);
+
+      // Trace back through the SSA chain to find the affine.load
+      if (auto definingOp = operand.getDefiningOp()) {
```

Comment:
:warning: **llvm\-qualified\-auto** :warning:
`` auto definingOp `` can be declared as `` auto *definingOp ``
```suggestion
      if (auto *definingOp = operand.getDefiningOp()) {
```